### PR TITLE
(DOCSP-25080) Add sync section to JS quickstarts

### PR DESCRIPTION
**Pull request by krollins-mdb**
## Pull Request Info

The new quickstart sync section for React Native and Node.js is the same, as they both use the same codebase. However, the React Native quickstart will be updated in the near future to use Realm React - so an include doesn't make sense here.

The new section is similar to the [Kotlin quickstart](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-25080/sdk/kotlin/quick-start/#add-device-sync--optional-).

### Jira

- https://jira.mongodb.org/browse/DOCSP-25080

### Staged Changes

- [Node.js quickstart](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-25080/sdk/node/quick-start/)
- [React Native quickstart](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-25080/sdk/react-native/quick-start/)

### Reminder Checklist

If your PR modifies the docs, you might need to also update some corresponding
pages. Check if completed or N/A.

- [ n/a ] Create Jira ticket for corresponding docs-app-services update(s), if any
- [x] Checked/updated Admin API
- [x] Checked/updated CLI reference

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/react-native/quick-start.txt [hunk 1]
```diff
@@ -1,437 +1,309 @@
...
+Open a Synced Realm
+~~~~~~~~~~~~~~~~~~~
+
+After you have initialized your App, authenticated a user, and
+defined your object model, you can create a :js-sdk:`SyncConfiguration
+<Realm.App.Sync.html#~SyncConfiguration>`:
+
+If you have opened a local realm following the :ref:`Open a Realm
+<react-native-quickstart-local-open-a-realm>` section above, replace the
+``RealmConfiguration `` with the ``SyncConfiguration`` described below.
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:265]:
```suggestion
``RealmConfiguration`` with the ``SyncConfiguration`` described below.
```

fixes rendering issue

File: source/sdk/react-native/quick-start.txt [hunk 2]
```diff
@@ -1,437 +1,309 @@
...
-         user.logOut();
+   // Returns the shared instance of the Realm app.
+   export function getRealmApp() {
+      const appId = '<enter your App ID here>'; // Set App ID here.
+      const appConfig = {
+        id: appId,
+        timeout: 10000,
+      };
+     return new Realm.App(appConfig);
+   }
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:227]: this is a _funky_ way of instantiating the Realm.App instance.

also i think the comment is wrong.

this is creating a _new_ instance of the app every time you call the function. 

i think it'll still work, but it's odd and wastes compute resources reinstantiating the object every time you call the function.

rather you could just do:

```js
const appId = '<enter your App ID here>'; // Set App ID here.
const appConfig = {
  id: appId,
  timeout: 10000
};
const app = new Realm.App(appConfig);
```

File: source/sdk/react-native/quick-start.txt [hunk 3]
```diff
@@ -1,437 +1,309 @@
...
+
+.. code-block:: javascript
+   :emphasize-lines: 8,9
+
+   import Realm from 'realm';    
+
+   async function anonymousLogin() {
+     let user;
+     try {
+       const app = new Realm.App(appConfig); // pass in the appConfig variable that you created earlier
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:245]: this line is also _funky_ b/c it uses the `appConfig` variable which was scoped to the `getRealmApp()` function defined above.

this example should be adjusted based on changes made to above code example

File: source/sdk/react-native/quick-start.txt [hunk 4]
```diff
@@ -1,437 +1,309 @@
...
+If you have opened a local realm following the :ref:`Open a Realm
+<react-native-quickstart-local-open-a-realm>` section above, replace the
+``RealmConfiguration `` with the ``SyncConfiguration`` described below.
+
+Pass the authenticated user and the ``Item`` class to the ``SyncConfiguration``
+function to create a :ref:`Flexible Sync Configuration
+<react-native-flexible-sync>`.
+
+You need at least one subscription before you can read from or write to the
+realm. Use ``initialSubscriptions`` to define the initial subscription set when
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:272]:
```suggestion
realm. Use ``Configuration.sync.initialSubscriptions`` to define the initial subscription set when
```

File: source/sdk/react-native/quick-start.txt [hunk 5]
```diff
@@ -1,437 +1,309 @@
...
+   :emphasize-lines: 8,9
+
+   const config = {
+     sync: {
+       user: app.currentUser,
+       flexible: true,
+       initialSubscriptions: {
+         update: (subs, realm) => {
+           subs.add(
+             realm.objects("Item").filtered("name == 'User's Items'")
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:291]: this query doesn't make particular sense b/c why would multiple `Items` have the name `'User's Items'`?

rather could be smthn like:

```js
realm.objects("Item").filtered("ownerId == $0", app.currentUser.id)
```

File: source/sdk/react-native/quick-start.txt [hunk 6]
```diff
@@ -1,437 +1,309 @@
...
+           );
+         },
+       },
+     },
+   };
+   const realm = await Realm.open(config);
+
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:298]: i think would be good to follow this example with at least a sentence saying that now can write to realm and data syncs automatically to Atlas.

also data matching the subscription query + backend query filters automatically syncs down to the device.

Comment (mongodben) [reply source/sdk/react-native/quick-start.txt:298]: swift sdk has some good text you can borrow here - https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-25080/sdk/swift/quick-start/#open-a-realm-1

File: source/sdk/react-native/quick-start.txt [hunk 7]
```diff
@@ -1,437 +1,309 @@
...
+   .. literalinclude:: /examples/generated/node/quickstart-local.snippet.create-realm-objects.js
+     :language: javascript
 
-      .. code-block:: javascript
+.. _react-native-quickstart-local-find-sort-and-filter-objects:
 
-         // Query realm for all instances of the "Task" type.
-         const tasks = realm.objects("Task");
+Find, Sort, and Filter Objects
+------------------------------
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:84]:
```suggestion
------------------------------

```

File: source/sdk/react-native/quick-start.txt [hunk 8]
```diff
@@ -1,437 +1,309 @@
...
+.. _react-native-quickstart-local-complete:
 
-             // Update UI in response to modified objects
-             changes.modifications.forEach(index => {
-               let modifiedTask = tasks[index];
-               // ...
-             });
-           }
+Complete Example
+----------------
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:160]: we've removed the Complete Example sections from the other SDKs. should do so here as well

File: source/sdk/react-native/quick-start.txt [hunk 9]
```diff
@@ -1,437 +1,309 @@
...
-           try {
-             const app = new Realm.App(appConfig); // pass in the appConfig variable that you created earlier
+To define a Realm object type, create a schema object that specifies the type's
+``name`` and ``properties``. The type name must be unique among object types in
+a realm.
 
-             const credentials = Realm.Credentials.anonymous(); // create an anonymous credential
-             user = await app.logIn(credentials);
-             return user;
+.. example::
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:41]: i recommend removing all `.. example::` directives on this page. that's generally what we've done across the docs. 

also seems like some examples on this page aren't in the directive as well.

File: source/sdk/react-native/quick-start.txt [hunk 10]
```diff
@@ -1,437 +1,309 @@
...
+transaction <react-native-write-transactions>`.
 
-.. tabs-realm-languages::
+.. example::
 
-   .. tab::
-      :tabid: javascript
+   In the following example, an application developer creates two tasks. One task
+   with a name of "go grocery shopping" and another with the name of "go
+   exercise". Both tasks have a status of "Open".
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:76]:
```suggestion
```

don't think this provides any value to reader

File: source/sdk/react-native/quick-start.txt [hunk 11]
```diff
@@ -1,437 +1,309 @@
...
+----------------------------------------------
+
+Check out the :ref:`template app <template-apps>` to experiment with another
+quick way to start coding with the Realm React Native SDK. The React Native SDK template,
+labeled ``react-native.todo.flex``, is a premade application that integrates
+Realm Database and Atlas Device Sync into a customizable android app.
+
+Alternatively, if you are interested in a guided experience, you can read our
+:ref:`Android with Kotlin SDK tutorial <react-native-tutorial>` that expands on
+the template app.
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:309]: merp, react native pls

**Review by mongodben — COMMENTED** [review]
nice start, back to you! 

just applied suggestions to RN, but same applies to Node of course

File: source/sdk/react-native/quick-start.txt [hunk 12]
```diff
@@ -1,437 +1,264 @@
...
+integrated into your app.
 
-   This page contains only the essential information that you need to set up an
-   Atlas App Services backend. If you prefer to follow a guided tutorial that
-   shows you step-by-step how to set up a working app, check out the :ref:`React
-   Native Tutorial <react-native-tutorial>` where you'll build a mobile task
-   tracker app.
-
-.. _react-native-quick-start-import:
+.. _react-native-quickstart-local-import-realm:
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:21]: another perhaps out of scope comment, but a good quick win. there should be some info abt installing the sdk on this page. could be as simple as a bit of text and link to install page. 

see flutter sdk quick start for example - https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-25080/sdk/flutter/quick-start/

File: source/sdk/react-native/quick-start.txt [hunk 13]
```diff
@@ -1,437 +1,264 @@
...
+Authenticate a User
+~~~~~~~~~~~~~~~~~~~
+
+To authenticate and log in a user, call :js-sdk:`App.logIn()
+<Realm.App.html#logIn>`. When :ref:`anonymous authentication
+<anonymous-authentication>` is enabled, users can immediately log into your app
+without providing any identifying information:
+
+.. code-block:: javascript
+   :emphasize-lines: 8,9
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:184]:
```suggestion
```

this doesn't make sense anymore

File: source/sdk/react-native/quick-start.txt [hunk 14]
```diff
@@ -1,437 +1,264 @@
...
+   const appId = "<yourAppId>"; // Set App ID here.
+   const appConfig = {
+     id: appId,
+     timeout: 10000
+   }; 
+
+   async function anonymousLogin() {
+     let user;
+     try {
+       const app = new Realm.App(appConfig); // pass in the appConfig variable that you created earlier
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:197]:
```suggestion
       // pass in the appConfig variable that you created earlier
       const app = new Realm.App(appConfig);
```

nit: line was long and getting cut off. now easier to read

File: source/sdk/react-native/quick-start.txt [hunk 15]
```diff
@@ -1,437 +1,264 @@
...
+   import Realm from "realm";
+
+   const appId = "<yourAppId>"; // Set App ID here.
+   const appConfig = {
+     id: appId,
+     timeout: 10000
+   }; 
+
+   async function anonymousLogin() {
+     let user;
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:195]: there's no reason why this `user` should be declared outside the try block b/c it's only used within it.

```suggestion
```

File: source/sdk/react-native/quick-start.txt [hunk 16]
```diff
@@ -1,437 +1,264 @@
...
+     try {
+       const app = new Realm.App(appConfig); // pass in the appConfig variable that you created earlier
+
+       const credentials = Realm.Credentials.anonymous(); // create an anonymous credential
+       user = await app.logIn(credentials);
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:200]:
```suggestion
       const user = await app.logIn(credentials);
```

File: source/sdk/react-native/quick-start.txt [hunk 17]
```diff
@@ -1,437 +1,264 @@
...
+         throw `Error logging in anonymously: ${JSON.stringify(error,null,2)}`;
+     }
+   }
+
+Open a Synced Realm
+~~~~~~~~~~~~~~~~~~~
+
+After you have initialized your App, authenticated a user, and
+defined your object model, you can create a :js-sdk:`SyncConfiguration
+<Realm.App.Sync.html#~SyncConfiguration>`:
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:213]:
```suggestion
<Realm.App.Sync.html#~SyncConfiguration>`.
```

File: source/sdk/react-native/quick-start.txt [hunk 18]
```diff
@@ -1,437 +1,264 @@
...
+realm. Use ``Configuration.sync.initialSubscriptions`` to define the initial subscription set when
+the Realm file is first opened. Pass the query you wish to subscribe to and a
+name for the subscription to the :js-sdk:`add()
+<Realm.App.Sync.MutableSubscriptionSet.html#add>` method on the
+``MutableSubscriptionSet``.
+
+.. code-block:: javascript
+   :emphasize-lines: 8,9
+
+   const config = {
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:233]: this config doesn't have the schema. 

i think you should just bluehawk the code examples here to reduce risk of error like this

**Review by mongodben — COMMENTED** [review]
nice improvements. few things remaining. 

and i think you might as well just bluehawk the whole page while you're at it. 

even if the react native quick start is completely refactored around Realm React, the bluehawkification will still be valid for the node quickstart.

File: source/sdk/react-native/quick-start.txt (hunk 12 above)
Comment (krollins-mdb) [reply source/sdk/react-native/quick-start.txt:21]: Good point. I'll include that in my next commit.

**Comment by krollins-mdb** [comment]
Thanks, @mongodben! I realized I was getting mixed up with the old local-only example and adding just a bit of flexible sync stuff in the new section. I created a new example file and Bluehawked it. 😄

File: source/sdk/react-native/quick-start.txt [hunk 19]
```diff
@@ -1,437 +1,220 @@
...
-               throw `Error logging in anonymously: ${JSON.stringify(error,null,2)}`;
-           }
-         }
-..
-   .. tab::
-      :tabid: typescript
-      code-block:: typescript
+The following code shows how to define an object model for a ``Task`` object. In this example:
+  
+- The ``primaryKey`` is the ``_id`` of type ``int`` (Another common type used for primary keys is :js-sdk:`ObjectId <Realm.Object.html#objectId>`).
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:45]:
```suggestion
- The ``primaryKey`` is the ``_id`` of type ``int``. (Another common type used for primary keys is :js-sdk:`ObjectId <Realm.Object.html#objectId>`.)
```

File: source/examples/generated/node/quickstart.snippet.open-realm-with-subscriptions.js [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    progressMinutes: "int?",
+    owner: "string?",
+    dueDate: "date?",
+  },
+  primaryKey: "_id",
+};
+
+// Create a `SyncConfiguration` object.
+const config = {
+  sync: {
```
Comment (mongodben) [inline source/examples/generated/node/quickstart.snippet.open-realm-with-subscriptions.js:26]: this doesn't have the `schema` property. you must have that to write data to the realm. see https://www.mongodb.com/docs/realm/sdk/node/examples/open-and-close-a-realm/#open-a-flexible-synced-realm

File: source/sdk/react-native/quick-start.txt [hunk 21]
```diff
@@ -1,437 +1,220 @@
...
 
-.. _react-native-quick-start-logout:
+Open a Synced Realm
+~~~~~~~~~~~~~~~~~~~
 
-Log Out
--------
+After you have initialized your App, authenticated a user, and
+defined your object model, you can create a :js-sdk:`SyncConfiguration
+<Realm.App.Sync.html#~SyncConfiguration>`.
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:195]: need to be clearer that `SyncConfiguration` is just part of the larger `Configuration` object. see the text at the beginning of this section - https://www.mongodb.com/docs/realm/sdk/node/examples/open-and-close-a-realm/#open-a-flexible-synced-realm

**Review by mongodben — COMMENTED** [review]
didn't do full review, but back to you on one important part abt the Configuration.sync property

File: source/sdk/react-native/quick-start.txt [hunk 22]
```diff
@@ -1,437 +1,227 @@
...
+a realm.
 
-             const credentials = Realm.Credentials.anonymous(); // create an anonymous credential
-             user = await app.logIn(credentials);
-             return user;
+The following code shows how to define an object model for a ``Task`` object. In this example:
+  
+- The ``primaryKey`` is the ``_id`` of type ``int``. (Another common type used for 
+  primary keys is :js-sdk:`ObjectId <Realm.Object.html#objectId>`.)
+- The ``name`` field is required. The
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:47]:
```suggestion
- The ``name`` field is required.
```

File: source/sdk/react-native/quick-start.txt [hunk 23]
```diff
@@ -1,437 +1,227 @@
...
+- The ``status`` is optional, denoted by the question mark immediately after the data type.
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:48]:
```suggestion
- The ``status`` and ``onwer_id`` fields are optional, denoted by the question mark immediately after the data type.
```

File: source/sdk/react-native/quick-start.txt [hunk 24]
```diff
@@ -1,437 +1,227 @@
...
+defined your object model, you can create a :js-sdk:`SyncConfiguration
+<Realm.App.Sync.html#~SyncConfiguration>`.
 
-Log Out
--------
+To open a Flexible Sync realm, call :js-sdk:`Realm.open() <Realm.html#.open>`. 
+Pass in a :js-sdk:`Configuration <Realm.html#~Configuration>`
+object, which must include the ``sync`` property defining a 
+:js-sdk:`SyncConfiguration <Realm.App.Sync.html#~SyncConfiguration>` object. 
+In the SyncConfiguration, you must include include a ``user`` and ``flexible:true``.
```
Comment (mongodben) [inline source/sdk/react-native/quick-start.txt:202]:
```suggestion
In the SyncConfiguration, you must include include a ``user`` and ``flexible: true``.
```

File: examples/node/Examples/quickstart.js [hunk 25]
```diff
@@ -162,3 +163,70 @@ describe("QuickStart Local", () => {
...
+        status: "string?",
+        progressMinutes: "int?",
+        owner: "string?",
+        dueDate: "date?",
+      },
+      primaryKey: "_id",
+    };
+    
+    // Create a `SyncConfiguration` object.
+    const config = {
```
Comment (mongodben) [inline examples/node/Examples/quickstart.js:205]:
```suggestion
    const config = {
      schema: [TaskSchema],
```

must include the schema in the config.

**Review by mongodben — COMMENTED** [review]
looking better! 

everything remaining small.

the thing about the `Configuration.schema` property being missing was still present from last review. added suggestion for adding. let me know if you have any questions for this.

File: examples/node/Examples/quickstart.js [hunk 26]
```diff
@@ -162,3 +163,70 @@ describe("QuickStart Local", () => {
...
+        user: app.currentUser,
+        // Set flexible sync to true to enable sync.
+        flexible: true,
+        // Define initial subscriptions to start syncing data as soon as the
+        // realm is opened.
+        initialSubscriptions: {
+          update: (subs, realm) => {
+            subs.add(
+              // Get objects that match your object model, then filter them
+              // by owner id.
```
Comment (mongodben) [inline examples/node/Examples/quickstart.js:217]:
```suggestion
              // by the `owner_id` queryable field.
```

File: source/sdk/react-native/quick-start.txt [hunk 27]
```diff
@@ -1,437 +1,227 @@
...
 
-Once logged in, you can log out:
+Additionally, you need at least one subscription before you can read from or write to the
+realm. Use ``Configuration.sync.initialSubscriptions`` to define the initial 
+subscription set when the Realm file is first opened.
 
-.. tabs-realm-languages::
+.. literalinclude:: /examples/generated/node/quickstart.snippet.open-realm-with-subscriptions.js
+  :language: javascript
```
[the remark this answers is not in the document]
Comment (krollins-mdb) [reply source/sdk/react-native/quick-start.txt:209]: Sorry for the misunderstanding! I've fixed this in my latest commit.

File: examples/node/Examples/quickstart.js (hunk 25 above)
Comment (krollins-mdb) [reply examples/node/Examples/quickstart.js:205]: Included this locally. Will push in my next commit.

**Review by mongodben — APPROVED** [review]
lgtm! nice work here kyle.

one small platform-related issue i noticed is that this text is wrapping weirdly: 

Screen Shot 2022-12-14 at 10 07 10 AM

of course not your fault as writer, but consider rephrasing to make the problem go away. (maybe just removing the parens?)

---

getting on the bandwagon:

image

**Pull request merged**